File: Golang/63.unique-path-ii.go
/*
 * @lc app=leetcode.cn id=63 lang=go
 *
 * [63] 不同路径 II
 */

package leetcode

// @lc code=start
func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	var m int = len(obstacleGrid)
	var n int = len(obstacleGrid[0])
	// initiate
	stepMap := make([][]int, m)
	for raw := 0; raw < m; raw += 1 {
		stepMap[raw] = make([]int, n)
	}

	// regressive
	for raw := 0; raw < m; raw += 1 {
		for col := 0; col < n; col += 1 {
			// met obstacle
			if obstacleGrid[raw][col] == 1 {
				stepMap[raw][col] = 0
			} else {
				// (0,0)
				if raw == 0 && col == 0 {
					stepMap[0][0] = 1
				} else if raw == 0 {
					stepMap[0][col] = stepMap[0][col-1]
				} else if col == 0 {
					stepMap[raw][0] = stepMap[raw-1][0]
				} else {
					// sum(left -> right, up -> down)
					stepMap[raw][col] = stepMap[raw-1][col] + stepMap[raw][col-1]
				}
			}
		}
	}

	return stepMap[m-1][n-1]
}

// @lc code=end

File: Cplusplus/63.unique-path-ii.cpp
/*
 * @lc app=leetcode.cn id=63 lang=cpp
 *
 * [63] 不同路径 II
 */

#include<iostream>
using namespace std;

// @lc code=start
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        // dynamic programming
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        vector<vector<int>> stepMap = vector<vector<int>>(m, vector<int>(n, 0));

        // regressive
        // record the number of path from (0,0) to (raw, col)
        for (int raw = 0; raw < m; ++raw) {
            for (int col = 0; col < n; ++col) {
                if (obstacleGrid[raw][col] == 1) {
                    // meet a obsticle, fail
                    stepMap[raw][col] = 0;
                } else {
                    if (raw == 0 && col == 0) {
                        stepMap[0][0] = 1;
                    } else if (raw == 0) {
                        // first raw
                        stepMap[0][col] = stepMap[0][col - 1];
                    } else if (col == 0) {
                        // first col
                        stepMap[raw][0] = stepMap[raw - 1][0]; 
                    } else {
                        // sum(left->right, up->down)
                        stepMap[raw][col] = stepMap[raw - 1][col] + stepMap[raw][col - 1];
                    }
                }
            }
        }
        
        return stepMap[m - 1][n - 1];
    }
};
// @lc code=end

